azuredevops.go

```diff
@@ -276,202 +160,388 @@ func runAzureDevOpsHandler(c *gin.Context) {
+			ProviderID:      provider.ProviderID,
+			GitURL:          gitURL})
 
-	bodyBytes, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		fmt.Println(err)
-		return []byte{}, err
+		fmt.Println("Unable to put project: ", err)
+		ginutil.WriteAPIClientWriteError(c, err,
+			fmt.Sprintf("Unable to import project '%s' from group '%s' at url '%s'.",
+				i.Project, i.Group, gitURL))
+		return wharfapi.Project{}, err
 	}
 
-	return bodyBytes, nil
+	return projectInDb, nil
 }
 
-func getProjectBranches(i importBody, project string) ([]azureDevOpsBranch, error) {
-	urlPath, err := url.Parse(i.URL)
+func postBranchesWritesError(c *gin.Context, client wharfapi.Client, i importBody,
+	project azureDevOpsProject, projectInDb wharfapi.Project) error {
+	repositories, err := getRepositoriesWritesError(c, i, project)
 	if err != nil {
-		fmt.Println("Unable to get url: ", err)
-		return []azureDevOpsBranch{}, err
+		fmt.Println("Unable to get repos: ", err)
+		return err
 	}
 
-	urlPath.Path = fmt.Sprintf("%v/%v/%v/%v/%v", i.Group, project, apiRepositories, project, refsPath)
-	q := url.Values{}
-	q.Add("api-version", "5.0")
-	q.Add("filter", "heads/")
-	urlPath.RawQuery = q.Encode()
+	projectBranches, err := getProjectBranchesWritesError(c, i, project.Name)
+	if err != nil {
+		fmt.Println("Unable to get project branches: ", err)
+		return err
+	}
 
-	fmt.Println(urlPath.String())
+	for _, branch := range projectBranches {
+		_, err := client.PutBranch(wharfapi.Branch{
+			Name:      branch.Name,
+			ProjectID: projectInDb.ProjectID,
+			Default:   branch.Ref == repositories.Value[0].DefaultBranch,
+			TokenID:   i.TokenID,
+		})
+		if err != nil {
+			fmt.Println("Unable to post branch: ", err)
+			ginutil.WriteAPIClientWriteError(c, err, fmt.Sprintf("Unable to import branch '%s'", branch.Name))
+			return err
+		}
+	}
 
-	body, err := getBodyFromRequest(i.User, i.Token, urlPath.String())
+	return nil
+}
 
-	projectRefs := struct {
-		Value []azureDevOpsRef `json:"value"`
-		Count int              `json:"count"`
-	}{}
+// getTokenByIDWritesError Writes an error to the gin Context when an error occurs!
+func getTokenByIDWritesError(c *gin.Context, client wharfapi.Client, tokenID uint) (wharfapi.Token, error) {
+	token, err := client.GetTokenById(tokenID)
+	if err != nil || token.TokenID == 0 {
+		fmt.Printf("Unable to get token. %+v", err)
+		ginutil.WriteAPIClientReadError(c, err,
+			fmt.Sprintf("Unable to get token by id %v.", tokenID))
+		return wharfapi.Token{}, err
+	}
 
-	err = json.Unmarshal(body, &projectRefs)
-	if err != nil {
-		fmt.Println("Unable to unmarshal refs: ", err)
-		return []azureDevOpsBranch{}, err
+	return token, nil
+}
+
+// getOrPostTokenWritesError Writes an error to the gin Context when an error occurs!
+func getOrPostTokenWritesError(c *gin.Context, client wharfapi.Client, i importBody) (wharfapi.Token, error) {
+	var err error
+	if i.User == "" && i.TokenID == 0 {
+		err = fmt.Errorf("both token and user were omitted")
```

Comment:
nit: Use `errors.New` when you're not doing any formatting

```suggestion
		err = errors.New("both token and user were omitted")
```